Add `mock.Protected().Verify` overloads with additional `exactParameterMatch` parameter

Shereef opened the pull request:
Closes #752.

stakx commented on src/Moq/Protected/IProtectedMock.cs:
```diff
@@ -140,6 +140,20 @@ public interface IProtectedMock<TMock> : IFluentInterface
+		/// <param name="exactParameterMatch">Should the parameter types match exactly types that were provided</param>
+		/// <param name="args">The optional arguments for the invocation. If argument matchers are used, 
+		/// remember to use <see cref="ItExpr"/> rather than <see cref="It"/>.</param>
+		void Verify(string methodName, Times times, bool exactParameterMatch, params object[] args);
```
Some thinking aloud below; no need for you to change anything.

If we placed `exactParameterMatch` before `times`, there would be less chance of a breaking change (in case where someone currently passes a `bool` as the first argument; see also the description of #751). On the other hand, it makes more sense to place all argument-related parameters next to each other instead of having them artificially separated by `times`.

The way you currently seems logical to me. And since Moq's `Protected`-APIs probably aren't the busiest ones we can probably risk introducing a breaking change here.

Shereef replied:
I initially put it before `times` but then I didn't like it and moved it to before `args`, I'd be more than happy to change it if you change your mind later on.

Shereef replied:
I think i need to resolve this conversation for the change request to be complete on GitHub, I maybe wrong, please unresolve it if you change your mind.

stakx commented on src/Moq/Protected/ProtectedMock.cs:
```diff
@@ -154,18 +154,20 @@ public ISetupSequentialResult<TResult> SetupSequence<TResult>(string methodOrPro
 		#region Verify
 
-		public void Verify(string methodName, Times times, object[] args)
+		public void Verify(string methodName, Times times, object[] args) { Verify(methodName, times, false, args); }
```
Nitpick: For consistency with other code, could you please reformat this to:

```diff
-public void Verify(string methodName, Times times, object[] args) { Verify(methodName, times, false, args); }
+public void Verify(string methodName, Times times, object[] args)
+{
+    this.Verify(methodName, times, false, args);
+}
+
```

Shereef replied:
Done

stakx commented on src/Moq/Protected/ProtectedMock.cs:
```diff
@@ -154,18 +154,20 @@ public ISetupSequentialResult<TResult> SetupSequence<TResult>(string methodOrPro
 		}
 
-		public void Verify<TResult>(string methodName, Times times, object[] args)
+		public void Verify<TResult>(string methodName, Times times, object[] args) { Verify<TResult>(methodName, times, false, args); }
```
Please reformat (as per the above review comment) for consistency. :)

Shereef replied:
Done

stakx commented on tests/Moq.Tests/VerifyFixture.cs:
```diff
@@ -6,6 +6,7 @@
 using Xunit;
 
 using System.Threading.Tasks;
+using Moq.Protected;
```
Nitpick: Please add one blank line above this one. In the Moq codebase, `using`s are generally grouped into blocks by their main namespace. `System.*` comes first, then any other groups follow in alphabetical order.

(`Xunit` should actually be placed after `System.*` but since that was already wrong before your PR, we'll just leave that as is for the moment.)

Shereef replied:
Done

stakx commented on tests/Moq.Tests/VerifyFixture.cs:
```diff
@@ -1569,6 +1581,27 @@ public interface ICat
 			string Purr(int amount);
 			void Hiss();
 		}
+		public class ParentDto { }
```
Please put additional blank lines between type definitions. :)

Shereef replied:
Done

stakx commented:
I hope you don't mind me having altered your PR's title and description.

* Regarding the changed title: Because GitHub will put the PR title into the merge commit message by default, and because the Moq source code might not live on GitHub for all eternity (at least in theory), it is good practice to make the PR titles independent of GitHub PR numbers and make them mostly self-explanatory.

* Regarding the added description: GitHub treats a few special words as instructions; notice the dotted underline. Putting the words _"fixes ..."_ or _"resolves ..."_ into the description means that when the PR gets merged, the referenced issue gets automatically closed.;-)

Shereef commented:
I wanted to know if there will be a Nuget build that has this soon, obviously, I would prefer it to be out sooner rather than later, however, I understand you may be holding off until v5 which is fine.

I have a workaround in place and wanted to know when should I plan on undoing this workaround. No rush. take you time.

stakx commented on CHANGELOG.md:
```diff
@@ -12,11 +12,12 @@ The format is loosely based on [Keep a Changelog](http://keepachangelog.com/en/1
 
 * All custom argument matcher methods (including those using `Match.Create<T>`) must now be marked with the `[Matcher]` attribute. For this reason, `MatcherAttribute` is no longer marked obsolete (@stakx, #732)
-* Method overload resolution for `mock.Protected().Setup("VoidMethod", ...)` may change due to a new overload: If you set up a `void` method whose first argument is a `bool`, make sure that argument gets interpreted as part of `args`, not as `exactParameterMatch` (see also *Added* section below). (@stakx, #751)
+* Method overload resolution for `mock.Protected().Setup("VoidMethod", ...)` and `mock.Protected().Verify(...)` may change due to a new overload: If you set up a `void` method whose first argument is a `bool`, make sure that argument gets interpreted as part of `args`, not as `exactParameterMatch` (see also *Added* section below). (@stakx & @Shereef, #751, #753)
```
Please note that the description is no longer quite correct. I'd suggest you fix this by rephrasing the following:

* We're talking about more than just one new method overload. Therefore:

   ```diff
   -due to a new overload:
   +due to new overloads:
   ```

* We're no longer taking just about setting up `void` methods; it's now also about verifying any type of method. Therefore let's be less specific:

   ```diff
   -If you set up a `void` method whose first argument is a `bool`,
   +If the first argument is a `bool`, 
   ```

Shereef replied:
Done, thanks for your attention to detail

stakx commented on CHANGELOG.md:
```diff
@@ -12,11 +12,12 @@ The format is loosely based on [Keep a Changelog](http://keepachangelog.com/en/1
 #### Added
 
 * New method overload `mock.Protected().Setup("VoidMethod", exactParameterMatch, args)` overload having a `bool exactParameterMatch` parameter. Due to method overload resolution, it was easy to think this already existed resolution when in fact it did not, leading to failing tests. (@stakx, #751)
+* New method overload `mock.Protected().Verify("VoidMethod", times, exactParameterMatch, args)` overload having a `bool exactParameterMatch` parameter. Due to method overload resolution. (@Shereef, #753)
```
Similar to the above comment, I'd suggest two small changes here:

* You added two overloads, not just one. Therefore (2x):

   ```diff
   -New method overload `mock.Protected().Verify("VoidMethod", times, exactParameterMatch, args)` overload
   +New method overloads `mock.Protected().Verify("VoidMethod", times, exactParameterMatch, args)`
   ```

* Drop the second sentence:

   ```diff
   -Due to method overload resolution.
   ```

Shereef replied:
Done

stakx commented:
> I can't get it to change the status from changes requested, Am I doing something wrong ?

Only the reviewer (and possibly other Moq team members) can change that status. So don't worry about it, you're doing the right thing. ;-)

(Btw. you can usually also leave the "resolve conversation" thingy to the reviewer who initiallycommented, or requested a change. It's a useful tool for them to track remaining changes.)

> I wanted to know if there will be a Nuget build that has this soon, obviously, I would prefer it to be out sooner rather than later

You already found the milestone. To put its remaining issues into context of a concrete time frame, know that one of them is a rather large internal re-architecting that I have been planning and preparing for more than a year... so the 4.11.0 release is likely to take another two to three months, I'm afraid.

If there's more delay than that, I'll consider doing an ad-hoc 4.10.2 release including the non-breaking changes from 4.11.0.

Shereef commented:
> You already found the milestone. To put its remaining issues into context of a concrete time frame, know that one of them is a rather large internal re-architecting that I have been planning and preparing for more than a year... so the 4.11.0 release is likely to take another two to three months, I'm afraid.
> 
> If there's more delay than that, I'll consider doing an ad-hoc 4.10.2 release including the non-breaking changes from 4.11.0.

Thanks for your response, there is no hurry, take your time, this can wait a few months!

> (Btw. you can usually also leave the "resolve conversation" thingy to the reviewer who initiallycommented, or requested a change. It's a useful tool for them to track remaining changes.)

Will do!


> you're doing the right thing. ;-)

Thanks for the assurance and the explanations!

Shereef commented:
I prefer it if we merge #755 first then I can fix the merge conflict and then this one so that the imports are looking good.

I am open to suggestions if you don't like #755 or want something else done.

I just want the parts I touch in the Test Fixture for Verify to be 100% up to par.

stakx commented:
I've noticed that you've rewritten your PR to a few focused commits... awesome work, I couldn't ask for more. 😍

> I prefer it if we merge #755 first then I can fix the merge conflict and then this one so that the imports are looking good.

I've merged #755. (Thanks again for that, btw.!)

Everything is looking good with this PR, but if you want to make any final changes (e.g. rebase it on top of `master`), you may do so now. Otherwise I'll go ahead with the merge in half a day or so.

Shereef commented:
> I've noticed that you've rewritten your PR to a few focused commits... awesome work, I couldn't ask for more. 😍

I found a secret PR about contribution ^_^, my pleasure to comply!

> I've merged #755. (Thanks again for that, btw.!)

I have rebased, fixed the first commit to comply with the grouping of imports then force pushed, it's really my pleasure! ;)

stakx commented:
All good! 🚀 Your contribution is very much appreciated, thank you!

Shereef commented:
> All good! 🚀 Your contribution is very much appreciated, thank you!
Your repository and all the work you do for it is much more appreciated, Thank you!

stakx commented:
Well, I'm actually just the maintainer, not the owner of this project & repository, but thanks for the kinds words nevertheless. :heart:

Shereef commented:
You are a maintainer that replies and works on 24/7 to/on this repository you may not really own it, but it is clear to me that you are someone who takes ownership of whatever project they are handling and treats it accordingly, so thank you!

Let me know if there is anything I can do to make this ship sooner to the public or just generally help.
